File: DS_basics_no_visualizations/tree/binary_tree/compute_depth.py
"""
下面的例子树是：
         1
        / \
       2   3
      / \
     4   5
    /
   6

"""


class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def max_depth(root, depth=1):
    print("-----" * depth)
    if not root:
        print("max_depth(None)开始递归")
        print("max_depth(None)结束，返回值为：0")
        print("-----" * depth)
        return 0
    print(f"max_depth({root.value})开始递归")
    left_depth = max_depth(root.left, depth + 1)
    right_depth = max_depth(root.right, depth + 1)
    result = max(left_depth, right_depth) + 1
    print(f"max_depth({root.value})结束，返回值为：{result}")
    print("-----" * depth)
    return result


# 示例用法
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)
root.left.left.left = Node(6)

print("计算二叉树深度：")
print("二叉树的深度为：", max_depth(root))

File: DS_basics_no_visualizations/tree/binary_tree/compute_depth.cpp
/*

下面的例子树是：
         1
        / \
       2   3
      / \
     4   5
    /
   6

*/
#include <iostream>
using namespace std;

struct Node
{
    int value;
    Node *left;
    Node *right;
    Node(int val) : value(val), left(NULL), right(NULL) {}
};

int maxDepth(Node *root, Node *parent = NULL, string direction = "root")
{
    if (root == NULL)
    {
        cout << "========================================" << endl;
        if (parent)
            cout << "maxDepth(NULL)开始递归，此为节点" << parent->value << "的" << direction << "子树" << endl;
        else
            cout << "maxDepth(NULL)开始递归" << endl;
        cout << "maxDepth(NULL)结束，返回值为：0" << endl;
        cout << "========================================" << endl;
        return 0;
    }
    cout << "========================================" << endl;
    if (parent)
        cout << "从节点" << parent->value << "进入其" << direction << "子节点" << root->value << endl;
    cout << "maxDepth(" << root->value << ")开始递归" << endl;
    cout << "----------------------------------------" << endl;
    int leftDepth = maxDepth(root->left, root, "左");
    int rightDepth = maxDepth(root->right, root, "右");
    cout << "节点" << root->value << "的左子树深度为：" << leftDepth << endl;
    cout << "节点" << root->value << "的右子树深度为：" << rightDepth << endl;
    int result = max(leftDepth, rightDepth) + 1;
    cout << "取其最大值，所以节点" << root->value << "的深度为：" << result << endl;
    cout << "maxDepth(" << root->value << ")结束，返回值为：" << result << endl;
    if (parent)
        cout << "此时从节点" << root->value << "返回节点" << parent->value << "，继续调用maxDepth" << endl;
    cout << "========================================" << endl;
    return result;
}

int main()
{
    Node *root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->left->left->left = new Node(6);

    cout << "计算二叉树深度：" << endl;
    cout << "二叉树的深度为：" << maxDepth(root) << endl;

    delete root->left->left->left;
    delete root->left->right;
    delete root->left->left;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}
